<a href="https://colab.research.google.com/github/RedStoneJeong/Data-Science/blob/main/Exercise_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 1)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Pandas Series


### Exercise 3: Analyzing Columns / Pandas Series

In [1]:
#run the cell!
import pandas as pd

In [3]:
#run the cell!
cars = pd.read_csv("cars.csv")

In [4]:
#run the cell!
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa             chevrolet s-10  

[398 rows x 9 columns]

In [5]:
#run the cell!
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


18. Select the numerical column __"mpg"__, create a __copy__ and __save__ the column/Pandas Series in the variable __mpg__! __Fill in the gaps__!

In [6]:
mpg = cars.mpg.copy()

In [7]:
#run the cell!
mpg

0      18.0
1      15.0
2      18.0
3      16.0
4      17.0
       ... 
393    27.0
394    44.0
395    32.0
396    28.0
397    31.0
Name: mpg, Length: 398, dtype: float64

19. Get some __summary statistics__ on the Series __mpg__! What is the __mpg__ of the __least fuel efficient__ car?

In [8]:
mpg.describe()

count    398.000000
mean      23.514573
std        7.815984
min        9.000000
25%       17.500000
50%       23.000000
75%       29.000000
max       46.600000
Name: mpg, dtype: float64

The mpg of the least fuel efficient car is ... 9.0 miles per gallon!

20. Get the __maximum Value__ in the Series __mpg__ by explicitly calling the ... method! The __most fuel efficient__ car has a __mpg__ of...?

The most fuel efficient car has a mpg of... 46.6!

21. Get the Frequency/__Counts__ of all unique __values__ in the Series __mpg__! What is the __most frequent__ value?

The most frequent value is ... 13.0!

22. Get the __relative frequencies__ in the Series __mpg__! What is the __relative frequency__ of the __most frequent value__?

The relative frequency of 13.0 is ... 0.050251!

23. __Sort__ the Series __mpg__ from __low to high__! What is the __second lowest__ value?

The second lowest value is ... 10.0!

24. __Sort__ the Series __mpg__ from __high to low__ and __save the changes__ by setting the __inplace__ parameter to __True__!  <br>
__Fill in the gaps!__

In [ ]:
mpg.sort_values(--- = ---, inplace = ---)

Inspect the __first 5 elements__ of mpg! What is the __second highest__ value?

In [ ]:
#run the cell!
mpg.head()

The second highest value is ... 44.6!

25. __Sort__ the Series __mpg__ by the __Index__ and __save__ the changes!

In [ ]:
#run the cell!
mpg

__Miles per Gallon__ (mpg) can be transformed into __Liter per 100 Kilometer__ with the following formula:<br>
__Liter per 100 Kilometer = 235.21 / mpg__<br><br>
26. Create a new Pandas Series __l_per_100__ by applying the above formula! __Round__ the results to __2 decimals__! __Fill in the gaps__!

In [ ]:
--- = (---/---).round(2)

Run and Inspect. What is the __very first__ element?

In [ ]:
#run the cell!
l_per_100

The first element is ... 13.07!

27. Get some __summary statistics__ on the Series __l_per_100__! What is the __average__ value?

The average value is approx. ... 11.21!

28. Select the non-numerical column __"origin"__, create a __copy__ and __save__ the column/Pandas Series in the variable __origin__! 

Inspect! The first 5 elements are all...?

In [ ]:
#run the cell!
origin

The first 5 elements are all ... usa!

29. Call the __describe()__ method on the non-numerical Series __"origin"__! What is the __most frequent__ value?

The most frequent value is ... usa!

30. Get all __unique values__ in the Series __origin__! Apart from the value usa, there are also the values...?

... japan and europe!

31. Last but not least, __count the frequencies__ in the Series __origin__! How often does the value __europe__ appear?

The value europe appears ... 70 times!

# Well Done!

------------------------------------------------------------

# Hints (Spoiler!)

18. mpg

19. describe() method

20. max() method

21. value_counts() method

22. normalize parameter

23. sort_values() method

24. ascending & inplace parameter

25. sort_index() method; inplace parameter

26. 235.21/mpg

27. describe() method

28. cars.---.---()

29. describe() method

30. unique() method

31. value_counts() method